elasticapm/contrib/serverless/__init__.py

```diff
@@ -0,0 +1,190 @@
+            elasticapm.instrument()
+
+    def __call__(self, func):
+        self.name = self.name or get_name_from_func(func)
+
+        @functools.wraps(func)
+        def decorated(*args, **kwds):
+            if len(args) == 2:
+                # Saving these for request context later
+                self.event, self.context = args
+            else:
+                self.event, self.context = {}, {}
+            if not self.client.config.debug and self.client.config.instrument:
+                with self:
+                    self.response = func(*args, **kwds)
+                    return self.response
+            else:
+                return func(*args, **kwds)
+
+        return decorated
+
+    def __enter__(self):
+        """
+        Transaction setup
+        """
+        trace_parent = TraceParent.from_headers(self.event.get("headers"))
+        self.transaction = self.client.begin_transaction("request", trace_parent=trace_parent)
+        elasticapm.set_context(
+            lambda: get_data_from_request(
+                self.event,
+                capture_body=self.client.config.capture_body in ("transactions", "all"),
+                capture_headers=self.client.config.capture_headers,
+            ),
+            "request",
+        )
+        if "httpMethod" in self.event and "resource" in self.event:
+            elasticapm.set_transaction_name("{} {}".format(self.event["httpMethod"], self.event["resource"]))
+        else:
+            elasticapm.set_transaction_name(self.name, override=False)
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        """
+        Transaction teardown
+        """
+        if exc_val:
+            self.client.capture_exception(exc_info=(exc_type, exc_val, exc_tb), handled=False)
+
+        if self.response and isinstance(self.response, dict):
+            elasticapm.set_context(
+                lambda: get_data_from_response(self.response, capture_headers=self.client.config.capture_headers),
+                "response",
+            )
+            if "statusCode" in self.response:
+                result = "HTTP {}xx".format(int(self.response["statusCode"]) // 100)
+                elasticapm.set_transaction_result(result, override=False)
+        self.client.end_transaction()
+
+
+def get_data_from_request(event, capture_body=False, capture_headers=True):
+    """
+    Capture context data from API gateway event
+    """
+    result = {}
+    if capture_headers and "headers" in event:
+        result["headers"] = event["headers"]
+    if capture_headers and "requestContext" in event:
+        result["requestContext"] = event["requestContext"]
+    if "httpMethod" not in event:
+        # Not API Gateway
+        return result
+
+    result["method"] = event["httpMethod"]
+    if event["httpMethod"] in constants.HTTP_WITH_BODY and "body" in event:
+        body = event["body"]
+        if event.get("isBase64Encoded"):
```

Comment:
is there a way to avoid base64 decoding / JSON loading if `capture_body` is set to False anyway?

Reply:
Definitely. I'll make that change.